### ISSUE TITLE
Fast fail

### ISSUE THREAD

**Issue by cospectrum** [model-checking/kani#3458]
Add a flag to exit after the first failed verification. Have not found a way to do this.
Use case: ci speed up.

**Comment by zhassan-aws** [comment]
Thanks for the feature request @cospectrum. This makes sense.

There is a pretty ugly workaround to achieve that currently, which is to run with:
```
--no-assertion-reach-checks --output-format old --enable-unstable --cbmc-args --stop-on-fail
```
but we should add a Kani option instead.

If you're willing to add an option, we can provide some guidance.

**Comment by zhassan-aws** [comment]
Stopping on failure can be on two different levels:
1. If there are multiple harnesses, stop as soon as one of the harnesses fail
2. Within one harness: if a check fails, do not continue with the remaining checks

The second level is harder to implement because with CBMC's `--stop-on-fail` option, it can stop because of a reachability check passing (i.e. an assertion is reachable).

**Comment by rajath-mk** [comment]
Hi @zhassan-aws , do we add the option to the common args (`kani-driver/src/args/common.rs`) or the verification args (`kani-driver/src/args/mod.rs`) ? 

To implement the fast fail mechanism itself, I am looking at the try_fold iterator (`library/core/src/iter/traits/iterator.rs`). 

As discussed, I am focusing on the implementation for the sequential part now, and in the later stages maybe ill look into the parallel step.

**Comment by zhassan-aws** [comment]
> Hi @zhassan-aws , do we add the option to the common args (kani-driver/src/args/common.rs) or the verification args (kani-driver/src/args/mod.rs) ?

This should be part of the verification args.

> To implement the fast fail mechanism itself, I am looking at the try_fold iterator (library/core/src/iter/traits/iterator.rs).

This is the loop that runs all harnesses:
https://github.com/model-checking/kani/blob/380b1fb096c393407134258c8ced2c46661449f6/kani-driver/src/harness_runner.rs#L40

**Comment by zhassan-aws** [comment]
> The parallel execution produces non-deterministic output ordering, making it difficult to write stable tests.

One idea is to have n harnesses that all fail, and run with a smaller number of jobs, e.g. 2. I would expect in this case that Kani would always print:
```
Complete - 0 successfully verified harnesses, 2 failures, 2 total.
```

**Comment by rajath-mk** [comment]
Yes, that works well. Thanks. I have made another commit with the changes here: https://github.com/rajath-mk/kani/commit/ac185792f7f7755049382948f13ced1e12e6552b

The current approach has minimal changes to the original code and has additional channels and an error flag. But in order to utilize Rayon's par_iter's built in functionalities, I am looking into the try_for_each (which works but we lose some features such as inter thread comm), the try-fold method is suiting our purposes well based on the method definition. Will look into that.

**Comment by rajath-mk** [comment]
As discussed, added custom error handler with index based fetching --> in an event of a failure, the current index of the failed harness in the vector is captured, and later the element is obtained --> giving us essentially the first failed harness among all threads.

code working as expected: https://github.com/rajath-mk/kani/commit/88e6bed003ca73a9b224e313b2a108df95417a06

@zhassan-aws
